### PR TITLE
Add fluent API for configuring owned entity types.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
@anpete, @ajcvickers, @rowanmiller, @divega See OwnedTypesTestBase.cs for some examples

### CODE REVIEW

File: src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 1]
```diff
@@ -194,6 +195,60 @@ public virtual IndexBuilder HasIndex([NotNull] params string[] propertyNames)
 
         /// <summary>
         ///     <para>
+        ///         Configures a relationship where this entity type provides identity to
+        ///         the other type in the relationship.
```
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder.cs]: Maybe this?

> Configures a relationship where the target entity is owned by (or part of) this entity. The target entity does not have it's own key. It's key value is always controlled by the entity is belongs to.

File: src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder`.cs [hunk 2]
```diff
@@ -179,6 +176,33 @@ public virtual IndexBuilder HasIndex([NotNull] Expression<Func<TEntity, object>>
...
         /// <summary>
         ///     <para>
+        ///         Configures a relationship where this entity type provides identity to
+        ///         the other type in the relationship.
+        ///     </para>
+        /// </summary>
+        /// <typeparam name="TRelatedEntity"> The entity type that this relationship targets. </typeparam>
+        /// <param name="navigationExpression">
+        ///     A lambda expression representing the reference navigation property on this entity type that represents
+        ///     the relationship (<c>post => post.Blog</c>).
```
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder`.cs]: I think the example should be representative of this type of relationship, so maybe `<c>customer => customer.Address</c>`

File: src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder.cs (hunk 1 above)
Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder.cs]: That's not accurate. The owned type can be the principal in the relationship between it and the owner type.
The owner, or more specifically the navigation on the owner, just provides a way of finding the related type.

**Comment by rowanmiller** [comment]
From an API perspective, this looks pretty good to me. We may come up with a better name for `Owns(...)`... but I don't think that is a bad name.

Left a few comments with doc suggestions

File: src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder.cs (hunk 1 above)
Comment (rowanmiller) [inline src/Microsoft.EntityFrameworkCore/Metadata/Builders/EntityTypeBuilder.cs]: Let's discuss in person 😄

**Comment by AndriySvyryd** [comment]
Updated with feedback from the design meeting

File: test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+using Xunit;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore.Tests
+{
+    public abstract partial class ModelBuilderTest
+    {
+        public abstract class OwnedTypesTestBase : ModelBuilderTestBase
+        {
+            //[Fact]
```
Comment (ajcvickers) [inline test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs]: > //[Fact] [](start = 11, length = 9)

Why are the tests commented out?

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 4]
```diff
@@ -35,13 +34,6 @@ public EntityTypeBuilder([NotNull] InternalEntityTypeBuilder builder)
             Builder = builder;
         }
 
-        /// <summary>
-        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
-        ///     directly from your code. This API may change or be removed in future releases.
-        /// </summary>
-        protected virtual EntityTypeBuilder New([NotNull] InternalEntityTypeBuilder builder)
-            => new EntityTypeBuilder(builder);
-
```
Comment (ajcvickers) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: Why did we used to have this and why do we not need it now?

File: test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs (hunk 3 above)
Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs]: They don't pass without changes to conventions (next PR)

Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs]: (The tests are still valuable as a compile-time check of the API)

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs (hunk 4 above)
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs]: It allowed to reuse the base implementation.
Now I realized that the public builders shouldn't have much implementation apart from calling into the internal builder, so reusing the same boilerplate code is not worth a virtual call and a cast

File: src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs [hunk 5]
```diff
@@ -135,28 +127,16 @@ public virtual PropertyBuilder<TProperty> Property<TProperty>([NotNull] Expressi
...
         /// <param name="propertyName"> The name of then property to be removed from the entity type. </param>
         public new virtual EntityTypeBuilder<TEntity> Ignore([NotNull] string propertyName)
-        {
-            Check.NotEmpty(propertyName, nameof(propertyName));
-
-            Builder.Ignore(propertyName, ConfigurationSource.Explicit);
-
-            return this;
-        }
+            => (EntityTypeBuilder<TEntity>)base.Ignore(propertyName);
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs]: null check gone.

Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs]: The base method has the check

File: test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            public virtual void Can_use_nested_closure()
+            {
+                var modelBuilder = CreateModelBuilder();
+                var model = modelBuilder.Model;
+
+                modelBuilder.Entity<Customer>().OwnsOne(
+                    c => c.Details,
+                    r => r.HasAnnotation("foo", "bar")
+                        .HasOwnershipAnnotation("bar", "foo"));
+
```
Comment (ajcvickers) [inline test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs]: Get the concept (HasAnnotation puts the annotation on the entity type, while HasOwnershipAnnotation puts it on the FK) but not sure it is the best API. Could consider one to be YAGI, or maybe also explicitly name the entity type one to HasTypeAnnotation. @divega @anpete Ideas?

Comment (divega) [inline test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs]: I tend to agree with the suggestion but honestly I feel like I need a refresher of the whole thing to have more context.

**Comment by tinchou** [comment]
Hey guys,

Would you mind sharing a quick description of what Owned Entity Types are? I can't quite figure it out from looking at [`OwnedTypesTestBase.cs`](https://github.com/aspnet/EntityFramework/blob/dev/test/EFCore.Tests/ModelBuilderTest/OwnedTypesTestBase.cs) or some of the other PRs.

Thanks!

**Comment by AndriySvyryd** [comment]
@tinchou In short this enables scenarios like:
```C#
class Customer
{
    public int Id { get; set; }
    public Address BillingAddress { get; set; }
    public Address ShippingAddress { get; set; }
}

class Address
{
    public string Street { get; set; }
    public string City { get; set; }
}
```
There are two concepts here:
* __Delegated Identity Entity Types__ are entity types that are mapped to the same CLR type, since they cannot be identified by their name or CLR type they can only be accessed through a navigation from another entity type, hence their identity is delegated. It is possible to have multiple navigations to a DIET so its identity is not unique. The primary key is usually the same as the defining foreign key and usually consists of shadow properties. 
* __Owned Entity Types__ are DIETs where the defining entity type is marked as the owner. There can only be one navigation to an owned entity type. The owner must be the principal of the defining relationship. The navigation to an owned entity type will be automatically included in the query.

**Comment by tinchou** [comment]
So how would the tables look like in that case? I'm thinking the following, but that wouldn't be any different from regular relationships.

Id | BillingAddressId | ShippingAddressId
---|-------------------|------
... | ... | ...

AddressId | Street | City
---|-------------------|------
... | ... | ...

Thanks!

**Comment by AndriySvyryd** [comment]
The `Customer` is the principal and the `Address` DIETs are mapped to different tables, so it would be:

`Customers`:

| Id |
|---|
|...|

`Customers_BillingAddresses`:

CustomerId | Street | City
---|-------------------|------
... | ... | ...

`Customers_ShippingAddresses`:

CustomerId | Street | City
---|-------------------|------
... | ... | ...

However when we implement table splitting then by convention this would be mapped to a single table:

`Customers`:

Id | BillingAddress_Street | BillingAddress_City | ShippingAddress_Street | ShippingAddress_City
---|-------------------|------|-------------------|------
... | ... | ... | ... | ...

**Comment by tinchou** [comment]
Awesome!

**Pull request merged**